perf(linter): disable lint rules by file type

DonIsaac opened the pull request:
### TL;DR

Added a `should_run` function to multiple lint rules to determine if a rule should be executed based on the source type. This change optimizes the linting process by avoiding unnecessary rule checks.

### What changed?

1. **New Method**: Introduced the `should_run` method in the `Rule` trait.
2. **Implementation**: Implemented the `should_run` method for various lint rules, particularly those related to React and TypeScript.
3. **Usage**: Updated the `Linter` to use the `should_run` method to filter rules before execution.
4. **Macro Update**: Modified the `declare_all_lint_rules` macro to incorporate the `should_run` method.

### How to test?

1. Run the linter on a project containing React and TypeScript files.
2. Verify that only relevant rules are executed based on the file type (e.g., JSX rules for React files).

### Why make this change?

This change improves the performance of the linter by ensuring that only applicable rules are run for a given file type, reducing unnecessary computation and potential false positives.

---

DonIsaac commented:
* **#4380** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4380?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen commented:
The summary feels AI generated 👀

mysteryven commented:
Found several repeat check can be removed: 

https://github.com/oxc-project/oxc/blob/a207923af1b87953e89ecbf0fe272cf7134b71ef/crates/oxc_linter/src/rules/typescript/no_this_alias.rs#L90-L92
https://github.com/oxc-project/oxc/blob/a207923af1b87953e89ecbf0fe272cf7134b71ef/crates/oxc_linter/src/rules/typescript/no_unsafe_declaration_merging.rs#L39-L41
https://github.com/oxc-project/oxc/blob/a207923af1b87953e89ecbf0fe272cf7134b71ef/crates/oxc_linter/src/rules/typescript/no_var_requires.rs#L37-L39

DonIsaac commented:
> The summary feels AI generated 👀 

It is 👀

DonIsaac committed:
perf(linter): disable lint rules by file type (#4380)

### TL;DR

Added a `should_run` function to multiple lint rules to determine if a rule should be executed based on the source type. This change optimizes the linting process by avoiding unnecessary rule checks.

### What changed?

1. **New Method**: Introduced the `should_run` method in the `Rule` trait.
2. **Implementation**: Implemented the `should_run` method for various lint rules, particularly those related to React and TypeScript.
3. **Usage**: Updated the `Linter` to use the `should_run` method to filter rules before execution.
4. **Macro Update**: Modified the `declare_all_lint_rules` macro to incorporate the `should_run` method.

### How to test?

1. Run the linter on a project containing React and TypeScript files.
2. Verify that only relevant rules are executed based on the file type (e.g., JSX rules for React files).

### Why make this change?

This change improves the performance of the linter by ensuring that only applicable rules are run for a given file type, reducing unnecessary computation and potential false positives.

---
